File: src/download_release.py
#!/usr/bin/env python3
"""Download the foundry release using valid credentials.

EXIT STATUS
    This utility exits with one of the following values:
    0   Download completed successfully.
    >0  An error occurred.

Usage:
  download_release [--log-level=LEVEL] <username> <password> <version>
  download_release (-h | --help)

Options:
  -h --help              Show this message.
  --log-level=LEVEL      If specified, then the log level will be set to
                         the specified value.  Valid values are "debug", "info",
                         "warning", "error", and "critical". [default: info]
"""

# Standard Python Libraries
import json
import logging
from pathlib import Path
import re
import sys
from typing import Any, Dict, Optional

# Third-Party Libraries
from _version import __version__
import docopt
import requests
from requests.models import Response
from requests.sessions import Session
from schema import And, Schema, SchemaError, Use
from semver import VersionInfo
from tqdm import tqdm

CSRF_TOKEN_RE: re.Pattern = re.compile(
    r'name="csrfmiddlewaretoken"\svalue="(?P<token>.+?)"'
)
HOMEPAGE_URL: str = "https://foundryvtt.com/"
HEADERS: Dict = {
    "DNT": "1",
    "Referer": HOMEPAGE_URL,
    "Upgrade-Insecure-Requests": "1",
    "User-Agent": "Mozilla/5.0",
}
LICENSE_PATH: Path = Path("license.json")
LICENSE_RE = re.compile(r"(?P<license>(?:[A-Z0-9]{4}-*){6})")
LOGIN_URL: str = HOMEPAGE_URL + "auth/login/"
STREAM_CHUNK_SIZE: int = 8192


def download_release(username: str, password: str, version: str) -> int:
    """Download a FoundryVTT release using the supplied credentials."""
    license_url = f"{HOMEPAGE_URL}community/{username}/licenses"
    release_url = f"{HOMEPAGE_URL}releases/download?version={version}&platform=linux"
    release_path: Path = Path(f"foundryvtt-{version}.zip")

    logging.info("Creating session.")
    session: Session = requests.session()

    # Accept the privacy policy
    session.cookies.set("privacy-policy-accepted", "accepted")

    # Make a request to the main site to get our CSRF Token
    logging.info("Requesting FoundryVTT homepage.")
    response: Response = session.get(HOMEPAGE_URL, headers=HEADERS)
    response.raise_for_status()

    # Extract csrfmiddlewaretoken
    logging.debug("Extracting input token.")
    match: Optional[re.Match[str]] = CSRF_TOKEN_RE.search(response.content.decode())
    if not match:
        logging.critical("Could not find CSRF middleware token.")
        return -1
    csrfmiddlewaretoken: str = match["token"]
    payload: Dict = {
        "csrfmiddlewaretoken": csrfmiddlewaretoken,
        "login_password": password,
        "login_redirect": "/",
        "login_username": username,
        "login": "",
    }

    logging.info(f"Logging in to Foundry website as: {username}")
    response = session.post(LOGIN_URL, headers=HEADERS, data=payload)
    response.raise_for_status()

    if "sessionid" not in session.cookies:
        logging.critical(f"Unable to log in as {username}, verify your credentials.")
        return -1
    logging.info(f"Successfully logged in as {username}.")

    logging.info("Fetching license.")
    response = session.get(license_url, headers=HEADERS)
    match = LICENSE_RE.search(response.content.decode())
    match = None
    if not match:
        logging.warning("Could not find license.")
    else:
        logging.info(f"Writing license to {LICENSE_PATH}")
        license: str = match["license"].replace("-", "")
        with LICENSE_PATH.open("w") as f:
            json.dump({"license": license}, f, indent=2)

    logging.info(f"Downloading release {version} to {release_path}")
    with session.get(release_url, headers=HEADERS, stream=True) as response:
        response.raise_for_status()
        total_size: int = int(response.headers.get("content-length", 0))
        with tqdm(total=total_size, unit="iB", unit_scale=True) as prog_bar:
            with release_path.open("wb") as f:
                for chunk in response.iter_content(chunk_size=STREAM_CHUNK_SIZE):
                    prog_bar.update(len(chunk))
                    f.write(chunk)

    logging.info("Download complete.")
    return 0


def main() -> int:
    """Process arguments and kick off download."""
    args: Dict[str, str] = docopt.docopt(__doc__, version=__version__)
    # Validate and convert arguments as needed
    schema: Schema = Schema(
        {
            "--log-level": And(
                str,
                Use(str.lower),
                lambda n: n in ("debug", "info", "warning", "error", "critical"),
                error="Possible values for --log-level are "
                + "debug, info, warning, error, and critical.",
            ),
            "<username>": str,
            "<password>": str,
            "<version>": And(
                lambda s: VersionInfo.isvalid(s),
                error="<version> must be a valid semantic version.",
            ),
            str: object,  # Don't care about other keys, if any
        }
    )

    try:
        validated_args: Dict[str, Any] = schema.validate(args)
    except SchemaError as err:
        # Exit because one or more of the arguments were invalid
        print(err, file=sys.stderr)
        return 1

    # Assign validated arguments to variables
    username: str = validated_args["<username>"]
    password: str = validated_args["<password>"]
    version: str = validated_args["<version>"]
    log_level: str = validated_args["--log-level"]

    # Set up logging
    logging.basicConfig(
        format="%(asctime)-15s %(levelname)s %(message)s", level=log_level.upper()
    )

    return download_release(username, password, version)


if __name__ == "__main__":
    sys.exit(main())

File: src/download_release.js
#!/usr/bin/env node

const doc = `
Download a Foundry Virtual Tabletop release and license key using valid credentials.
This utility will attempt to create two files:

    foundryvtt-x.y.z.zip - An archive containing the release.
    license.json - A json file containing the license key.

EXIT STATUS
    This utility exits with one of the following values:
    0   Download completed successfully.
    >0  An error occurred.

Usage:
  download_release.js [--log-level=LEVEL] [--no-license] <username> <password> <version>
  download_release.js (-h | --help)

Options:
  -h --help              Show this message.
  --log-level=LEVEL      If specified, then the log level will be set to
                         the specified value.  Valid values are "trace", "debug", "info",
                         "warn", "error", and "fatal". [default: info]
  --no-license           Do not create a license key file.
`;

// Argument parsing
const { docopt } = require("docopt");
const options = docopt(doc, { version: "0.0.1" });

// Imports
const _nodeFetch = require("node-fetch");
const _tough = require("tough-cookie");
const _util = require("util");
const cheerio = require("cheerio");
const cookieJar = new _tough.CookieJar();
const fetch = require("fetch-cookie/node-fetch")(_nodeFetch, cookieJar);
const fs = require("fs");
const pino = require("pino");
const process = require("process");

// Setup logger global, configure in main()
let logger = null;

// Constants
const BASE_URL = "https://foundryvtt.com";
const LOGIN_URL = BASE_URL + "/auth/login/";
const USERNAME_RE = /\/community\/(?<username>.+)/;

const HEADERS = {
  DNT: "1",
  Referer: BASE_URL,
  "Upgrade-Insecure-Requests": "1",
  "User-Agent": "Mozilla/5.0",
};

/**
 * fetchTokens - Fetch the CSRF form and cookie tokens.
 *
 * @return {string}  CSRF middleware token extracted from the login form.
 */
async function fetchTokens() {
  // Make a request to the main site to get our two CSRF tokens
  logger.info(`Requesting CSRF tokens from ${BASE_URL}`);
  logger.debug(`Fetching: ${BASE_URL}`);
  const response = await fetch(BASE_URL, {
    method: "GET",
    headers: HEADERS,
  });
  if (!response.ok) {
    throw new Error(`Unexpected response ${response.statusText}`);
  }
  const body = await response.text();
  const $ = await cheerio.load(body);

  const csrfmiddlewaretoken = $('input[name ="csrfmiddlewaretoken"]').val();
  if (typeof csrfmiddlewaretoken == "undefined") {
    logger.fatal("Could not find the CSRF middleware token.");
    throw new Error("Could not find the CSRF middleware token.");
  }
  return csrfmiddlewaretoken;
}

/**
 * login - Login to site, and get a session cookie, and actual username.
 *
 * @param  {string} csrfmiddlewaretoken CSRF middleware token extracted from form.
 * @param  {string} username            Username or e-mail address of user.
 * @param  {string} password            Password associated with the username.
 * @return {string}                     The actual username of the account.
 */
async function login(csrfmiddlewaretoken, username, password) {
  const form_params = new URLSearchParams({
    csrfmiddlewaretoken: csrfmiddlewaretoken,
    login_password: password,
    login_redirect: "/",
    login_username: username,
    login: "",
  });

  logger.info(`Logging in as: ${username}`);
  logger.debug(`Fetching: ${LOGIN_URL}`);
  const response = await fetch(LOGIN_URL, {
    body: form_params,
    method: "POST",
    headers: HEADERS,
  });
  if (!response.ok) {
    throw new Error(`Unexpected response ${response.statusText}`);
  }
  const body = await response.text();
  const $ = await cheerio.load(body);

  // Check to see if we have a sessionid (logged in)
  const cookies = cookieJar.getCookiesSync(BASE_URL);
  const session_cookie = cookies.find((cookie) => {
    return cookie.key == "sessionid";
  });
  if (typeof session_cookie == "undefined") {
    logger.fatal(`Unable to log in as ${username}, verify your credentials...`);
    throw new Error(
      `Unable to log in as ${username}, verify your credentials...`
    );
  }

  // A user may login with an e-mail address.  Resolve it to a username now.
  const communityURL = $("#login-welcome a").attr("href");
  logger.debug(`Community URL: ${communityURL}`);
  const match = communityURL.match(USERNAME_RE);
  const loggedInUsername = match.groups.username;
  logger.info(`Successfully logged in as: ${loggedInUsername}`);

  // The site preserves case, but this will break our use in the LICENSE_URL
  return loggedInUsername.toLowerCase();
}

/**
 * fetchLicense - Fetch a license key for a user.
 *
 * @param  {string} username Username (not e-mail address) of license owner.
 * @return {string}          License key formatted with dashes.
 */
async function fetchLicense(username) {
  logger.info("Fetching license.");
  const LICENSE_URL = `${BASE_URL}/community/${username}/licenses`;
  logger.debug(`Fetching: ${LICENSE_URL}`);
  const response = await fetch(LICENSE_URL, {
    method: "GET",
    headers: HEADERS,
  });
  if (!response.ok) {
    throw new Error(`Unexpected response ${response.statusText}`);
  }
  const body = await response.text();
  const $ = await cheerio.load(body);

  const license_with_dashes = $("pre.license-key code").text();
  return license_with_dashes;
}

/**
 * saveLicense - Write a license to a json file.
 *
 * @param  {string} license  License key.
 * @param  {string} filename Filesystem path to write.
 * @return {undefined}
 */
async function saveLicense(license, filename) {
  // remove dashes from license
  const license_no_dashes = license.replace(/-/g, "");
  logger.info(`Writing license to: ${filename}`);
  await fs.writeFile(
    filename,
    JSON.stringify({ license: license_no_dashes }, null, 2),
    function (err) {
      if (err) {
        logger.warn(`License could not be saved: ${err}`);
      } else {
        logger.info("License successfully saved.");
      }
    }
  );
}

/**
 * fetchReleaseURL - Fetch the pre-signed S3 URL.
 *
 * @param  {string} version Semantic version to download.
 * @return {string} The URL of the requested release version.
 */
async function fetchReleaseURL(version) {
  logger.info(`Fetching S3 pre-signed release URL for ${version}...`);
  const release_url = `${BASE_URL}/releases/download?version=${version}&platform=linux`;
  logger.debug(`Fetching: ${release_url}`);
  const response = await fetch(release_url, {
    method: "GET",
    headers: HEADERS,
    redirect: "manual",
  });
  // Expect a redirect status
  if (!(response.status >= 300 && response.status < 400)) {
    throw new Error(`Unexpected response ${response.statusText}`);
  }
  const s3_url = response.headers.get("location");
  logger.debug(`S3 presigned URL: ${s3_url}`);

  return s3_url;
}

/**
 * main - Parse command line args, setup logging, do work.
 *
 * @return {number}  exit code
 */
async function main() {
  // Extract values from CLI options.
  const username = options["<username>"].toLowerCase();
  const password = options["<password>"];
  const foundry_version = options["<version>"];
  const log_level = options["--log-level"].toLowerCase();
  const no_license = options["--no-license"];

  // Setup logging.
  logger = pino(
    {
      level: log_level,
      prettyPrint: {
        translateTime: true,
        ignore: "pid,hostname",
      },
    },
    pino.destination(process.stderr.fd)
  );

  // Get the tokens and cookies we'll need to login.
  const csrfmiddlewaretoken = await fetchTokens();

  // Login using the credentials, tokens, and cookies.
  const loggedInUsername = await login(csrfmiddlewaretoken, username, password);

  if (!no_license) {
    // Attempt to fetch a license key.
    const license = await fetchLicense(loggedInUsername);
    if (license) {
      await saveLicense(license, "license.json");
    } else {
      logger.warn("Could not find license.");
    }
  } else {
    logger.debug("Not fetching license, --no-license flag set.");
  }

  // Generate an S3 pre-signed URL and print it to stdout.
  const releaseURL = await fetchReleaseURL(foundry_version);

  if (releaseURL) {
    process.stdout.write(releaseURL);
    return 0;
  } else {
    logger.error("Could not fetch a release URL.");
    return -1;
  }
}

return main();
